### PR TITLE
[Issue 9772][Go Functions]Allow User Metrics

### LINKED ISSUE #9772: [Go Functions] Allow user metrics
Currently Go function metrics are implemented with a Prometheus registry and [instantiated at runtime in an init()](https://github.com/apache/pulsar/blob/206ea82e012af34c5bb7f824826205019f3dd4ff/pulsar-function-go/pf/stats.go#L57-L150). These metrics are exposed to the functions runtime and automatically updated by the SDK, which makes sense, but the registry is inaccessible to the user.

It would be great to be able to support something like `recordMetric` as [provided to the Java SDK](https://pulsar.apache.org/docs/en/2.4.0/functions-metrics/)

A simple implementation could be to mirror this and provide a `context.RecordMetric(metricname string, value float64)`, though we could also provide package-level functions to create and register new metrics by name, like `pf.NewGauge(name string) (prometheus.Gauge, error)` which the user would then be responsible for updating in the code of their pulsar function.

### PR DESCRIPTION
**Pull request by flowchartsman**
Fixes #9772

### Motivation
#9772

### Modifications

Add context `RecordMetric(metricName string, metricValue float64)`

### Verifying this change

The current testing framework doesn't exercise pfunc metrics at all. If there are integration tests elsewhere that can verify this, let me know, but I have tested it in my environment and it works as intended.

### Does this pull request potentially affect one of the following parts:
  - Dependencies (does it add or upgrade a dependency): no
  - The public API: **yes** (pulsar function SDK only)
    - It brings the context object in-line with the other two frameworks, however it follows the same signature.
  - The schema: no
  - The default values of configurations: no
  - The wire protocol: no
  - The rest endpoints: no
  - The admin cli options: no
  - Anything that affects deployment: no

### Documentation

Check the box below and label this PR (if you have committer privilege).

Need to update docs? 

- [X] doc 

EDIT: I have removed the erroneous doc-required checkbox. This PR includes documentation.

### CODE REVIEW

File: site2/docs/functions-develop.md
Comment (Anonymitaet) [inline site2/docs/functions-develop.md:1009]: @freeznet could you please help review this from the technical perspective? Thanks

Comment (flowchartsman) [reply site2/docs/functions-develop.md:1009]: Worth noting that I more or less copied this from the Java one, since the method signature is the same and they use the same technique of overloading a single metric with labels.

File: pulsar-function-go/pf/instance.go
Comment (freeznet) [inline pulsar-function-go/pf/instance.go:90]: Can we move this part to `context.go`?

File: pulsar-function-go/pf/stats.go
Comment (freeznet) [inline pulsar-function-go/pf/stats.go:133]: please make sure the go's metrics are same as the java's
https://github.com/apache/pulsar/blob/3231caa894a24fa9048adf4628316d280d02c679/pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/ContextImpl.java#L204-L214

**Review by freeznet — CHANGES_REQUESTED** [review]
Thanks @flowchartsman for adding the user defined metrics to pulsar-functions-go, the impl LGTM, but I request changes on some details, please check the comments.
Besides the comments, please also remember to export the custom metrics in instance's `getMetrics()` so we can get the metrics with pulsar-admin. And please add some tests, thanks.

**Comment by flowchartsman** [comment]
Comment requested on latest batch of changes. Tests will drop when I figure out where to put them.

**Review by freeznet — DISMISSED** [review]
Great work, LGTM.

**Comment by freeznet** [comment]
@nlu90 @wolfstudy PTAL when you have time, thanks.

**Review by nlu90 — DISMISSED** [review]
LGTM

**Comment by flowchartsman** [comment]
Cool! Is anything further from me required to get this merged?

**Comment by flowchartsman** [comment]
Is anything blocking this from being merged?

**Comment by nlu90** [comment]
@flowchartsman

We need to 1. clear the CI checks, and 2. get one more Approve from another Committer.

@zymap or @wolfstudy Could you help take a look?

**Comment by flowchartsman** [comment]
If there's something I need to fix, could we get another test run, please? I cannot access the logs.

**Comment by flowchartsman** [comment]
Style failure looks like a network issue not (yet) a style one:

```
go: golang.org/x/tools@v0.0.0-20180917221912-90fa682c2a6e: git fetch -f origin refs/heads/*:refs/heads/* refs/tags/*:refs/tags/* in /home/runner/go/pkg/mod/cache/vcs/b44680b3c3708a854d4c89f55aedda0b223beb8d9e30fba969cefb5bd9c1e843: exit status 128:
	error: 9712 bytes of body are still expected
	fetch-pack: unexpected disconnect while reading sideband packet
	fatal: early EOF
	fatal: fetch-pack: invalid index-pack output
```

Especially since go1.13 style check passed. Could we get a re-run?

**Review by addisonj — APPROVED** [review]

**Review by freeznet — APPROVED** [review]

**Review by nlu90 — APPROVED** [review]

**Pull request merged**
